src/cache.ts

```diff
@@ -0,0 +1,135 @@
+/**
+ * @fileoverview this file provides methods handling dependency cache
+ */
+
+import { join } from 'path';
+import os from 'os';
+import * as cache from '@actions/cache';
+import * as core from '@actions/core';
+import * as glob from '@actions/glob';
+
+const STATE_CACHE_PRIMARY_KEY = 'cache-primary-key';
+const CACHE_MATCHED_KEY = 'cache-matched-key';
+const CACHE_KEY_PREFIX = 'setup-java';
+
+interface PackageManager {
+  id: 'maven' | 'gradle';
+  /**
+   * Paths of the file that specify the files to cache.
+   */
+  path: string[];
+  pattern: string[];
+}
+const supportedPackageManager: PackageManager[] = [
+  {
+    id: 'maven',
+    path: [join(os.homedir(), '.m2', 'repository')],
+    // https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#java---maven
+    pattern: ['**/pom.xml']
+  },
+  {
+    id: 'gradle',
+    path: [join(os.homedir(), '.gradle', 'caches'), join(os.homedir(), '.gradle', 'wrapper')],
+    // https://github.com/actions/cache/blob/0638051e9af2c23d10bb70fa9beffcad6cff9ce3/examples.md#java---gradle
+    pattern: ['**/*.gradle*', '**/gradle-wrapper.properties']
+  }
+];
+
+function findPackageManager(id: string): PackageManager {
+  const packageManager = supportedPackageManager.find(packageManager => packageManager.id === id);
+  if (packageManager === undefined) {
+    throw new Error(`unknown package manager specified: ${id}`);
+  }
+  return packageManager;
+}
+
+/**
+ * A function that generates a cache key to use.
+ * Format of the generated key will be "${{ platform }}-${{ id }}-${{ fileHash }}"".
+ * If there is no file matched to {@link PackageManager.path}, the generated key ends with a dash (-).
+ * @see {@link https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key|spec of cache key}
+ */
+async function computeCacheKey(packageManager: PackageManager) {
+  const hash = await glob.hashFiles(packageManager.pattern.join('\n'));
+  return `${CACHE_KEY_PREFIX}-${process.env['RUNNER_OS']}-${packageManager.id}-${hash}`;
+}
+
+/**
+ * Restore the dependency cache
+ * @param id ID of the package manager, should be "maven" or "gradle"
+ */
+export async function restore(id: string) {
+  const packageManager = findPackageManager(id);
+  const primaryKey = await computeCacheKey(packageManager);
+
+  core.debug(`primary key is ${primaryKey}`);
+  core.saveState(STATE_CACHE_PRIMARY_KEY, primaryKey);
+  if (primaryKey.endsWith('-')) {
+    core.warning(
```

Comment:
I wonder if we would like to `throw new Error()` if no files are found matched patterns like we have done in setup-node.

@AlenaSviridenko @MaksimZhukov  what do you think about it?

If we would like to keep it as a warning, I think we will need one more condition in `save` function because current implementation will try to save cache even in case when no files are matched patterns. I guess we don't want it.

Reply:
Thanks for your review! I just kept the existing behaviour (`@actions/cache` with `hashFiles(...)`), but it would be a great improvement making cache more intuitive. :+1:

I'll wait for the reply from other experts, and change the code once we made a consensus.

Reply:
+ @konradpabjan for any thoughts

Reply:
I made a proposal in my repo: https://github.com/KengoTODA/setup-java/commit/530d597222235536edde1012c73bc9034f8a1f97

If it looks good to you, I'll merge this commit to this PR. :)

Reply:
@KengoTODA , Proposal looks good to me. Please merge PR.

Also, I see that two checks fail right now:
- license - please ignore this check for now. I will fix it in the separate PR.
- build fails on verify_no_unstaged_changes - may be you haven't rebuild it last time? Let's try to rebuild one more time after merging your proposal via `npm run build && npm run release`